tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -201,6 +216,84 @@ private int Serialize(ref byte[] bytes, int offset, in SpanModel spanModel)
+            foreach (var spanLink in spanModel.Span.SpanLinks)
+            {
+                var context = spanLink.Context;
+                var samplingPriority = context.TraceContext?.SamplingPriority ?? context.SamplingPriority;
+                var traceFlags = samplingPriority switch
+                {
+                    null => 0u,             // not set
+                    > 0 => 1u + (1u << 31), // keep
+                    <= 0 => 1u << 31,       // drop
+                };
+                var len = 3;
+                // check to serialize tracestate
+                if (context.IsRemote)
+                {
+                    len++;
+                }
+
+                if (traceFlags > 0)
+                {
+                    len++;
+                }
+
+                var hasAttributes = spanLink.Attributes is { Count: > 0 };
+                if (hasAttributes)
+                {
+                    len++;
+                }
+
+                offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, len);
+                // individual key-value pairs - traceid - lower
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _traceIdBytes);
+                offset += MessagePackBinary.WriteUInt64(ref bytes, offset, context.TraceId128.Lower);
+                // individual key-value pairs - traceid - higher
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _traceIdHighBytes);
+                offset += MessagePackBinary.WriteUInt64(ref bytes, offset, context.TraceId128.Upper);
+                // spanid
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _spanIdBytes);
+                offset += MessagePackBinary.WriteUInt64(ref bytes, offset, context.SpanId);
+                // optional serialization
+                if (hasAttributes)
+                {
+                    offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _attributesBytes);
+                    offset += MessagePackBinary.WriteMapHeader(ref bytes, offset, spanLink.Attributes.Count);
+                    foreach (var attribute in spanLink.Attributes)
+                    {
+                        offset += MessagePackBinary.WriteString(ref bytes, offset, attribute.Key);
+                        offset += MessagePackBinary.WriteString(ref bytes, offset, attribute.Value);
+                    }
+                }
+
+                // CreateTraceStateHeader will never return null or empty
+                if (context.IsRemote)
+                {
+                    var traceState = W3CTraceContextPropagator.CreateTraceStateHeader(context);
+                    offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _traceStateBytes);
+                    offset += MessagePackBinary.WriteString(ref bytes, offset, traceState);
+                }
+
+                if (traceFlags > 0)
```

Comment:
From my understanding, I think this should be checking whether `traceFlags is not null` as if it is `0` we wouldn't currently serialize the value and from the RFC it seems that the keep/drop values should be included and it should only be omitted if the sampling isn't known.

Copying this comment to the other location as well.

Reply:
From the RFC (consolidating info here so I don't need to tab switch): 
`This field must be omitted if no sampling information was provided. If the original OpenTelemetry flags or the Datadog sampling priority are known, they should be directly provided here. If only sampling information is provided, the flags value must be 1 if the decision is keep, otherwise 0.`

and 
`    uint32 flags = 6;                           // Optional. W3C trace flags. If set, the high bit (bit 31) must be set.`

```
                var traceFlags = samplingPriority switch
                {
                    null => 0u,             // not set
                    > 0 => 1u + (1u << 31), // keep
                    <= 0 => 1u << 31,       // drop
                };
```

I think because we do the assignment of traceFlags in such a way that 0u would indicate that the traceFlags were never set, compared to when it is actually is 0, where we would set the highest bet. A scenario when the traceFlag should be 0, it would not actually be 0.

Hopefully I was able to explain myself properly, please let me know if I need to make things clearer.

Reply:
Thanks for the explanation I misunderstood what the RFC was asking and what we were doing 👍 

It's been a while since I've done bit manipulation. I also didn't realize that the backend (or something else) couldn't discern from a `0` or `null` value 

Your implementation is correct :smile:

Reply:
(replies in the other thread)